zerver/tests/test_email_notifications.py

```diff
@@ -1101,21 +1101,56 @@ def test_user_group_over_stream_wildcard_mention_priority(self) -> None:
         for text in expected_email_include:
             self.assertIn(text, self.normalize_string(mail.outbox[0].body))
 
-    def test_stream_wildcard_over_stream_mention_priority(self) -> None:
+    def test_followed_topic_stream_wildcard_over_stream_wildcard_mention_priority(self) -> None:
         hamlet = self.example_user("hamlet")
         othello = self.example_user("othello")
 
-        stream_mentioned_message_id = self.send_stream_message(othello, "Denmark", "1")
         stream_wildcard_mentioned_message_id = self.send_stream_message(
             othello, "Denmark", "@**all**"
         )
+        followed_topic_stream_wildcard_mentioned_message_id = self.send_stream_message(
+            othello, "Denmark", "@**all**"
+        )
 
         handle_missedmessage_emails(
             hamlet.id,
             [
                 {
-                    "message_id": stream_mentioned_message_id,
-                    "trigger": "stream_email_notify",
+                    "message_id": stream_wildcard_mentioned_message_id,
+                    "trigger": "stream_wildcard_mentioned",
+                    "mentioned_user_group_id": None,
+                },
+                {
+                    "message_id": followed_topic_stream_wildcard_mentioned_message_id,
+                    "trigger": "followed_topic_stream_wildcard_mentioned",
+                    "mentioned_user_group_id": None,
+                },
+            ],
+        )
+
+        expected_email_include = [
+            "Othello, the Moor of Venice: > @**all** > @**all** -- ",
+            "You are receiving this because you have wildcard mention notifications enabled for topics you follow.",
+        ]
+
+        for text in expected_email_include:
+            self.assertIn(text, self.normalize_string(mail.outbox[0].body))
+
+    def test_stream_wildcard_over_followed_topic_mention_priority(self) -> None:
+        hamlet = self.example_user("hamlet")
+        othello = self.example_user("othello")
+
+        followed_topic_mentioned_message_id = self.send_stream_message(othello, "Denmark", "1")
+        stream_wildcard_mentioned_message_id = self.send_stream_message(
+            othello, "Denmark", "@**all**"
+        )
+
+        handle_missedmessage_emails(
+            hamlet.id,
+            [
+                {
+                    "message_id": followed_topic_mentioned_message_id,
+                    "trigger": "followed_topic_email_notify",
```

Comment:
* We still have n assertions for n possible triggers
* I think it's better to test by sending separate messages having different elements (personal mention / wildcards), instead of a single message having all those elements. The reason being: it's more practical that a user has a few different missed messages having different elements and the email text priority is decided based on triggers of all those messages. a single message having all those triggers doesn't seems a practical case (although it will also verify the code logic).
* Having separate tests seems more readable to me.

Reply:
Cool, makes sense to me :)

Reply:
It seems the tests are patterned as `assert(P1>P2)`, `assert(P2>P3)`, ... where `Px` is priority `x`. So `n-1` tests in total :)
Might be helpful to add a comment here explaining the testing pattern, but not a blocker.
Also, I noticed that we currently have an existing test called `test_wildcard_over_stream_mention_priority`, the name of which is kinda misleading. `test_wildcard_mention_over_stream_message_priority` might be better?

Reply:
[#25978 (comment)](https://github.com/zulip/zulip/pull/25978#discussion_r1247709379)